### DISCUSSION TITLE
Terrible NFS performance and crashing parts of longhorn under heavy load

### DISCUSSION THREAD

**Discussion by linuxfreakus** [longhorn/longhorn#8111] [General]
I'm currently running Longhorn 1.5.3 and I have a large 120Ti RWX volume.  It is getting absolutely awful performance.  The underlying storage is a group of spinning sata grouped together with LVM.  It struggles to achieve 100MB/s of write speed at best, and seems to hang out more like 50 - 80 MB/s sometimes as low as 10 MB/s

This behavior is occurring while only ONE pod is mounting it and simply running rsync to copy 1GB data files (in one direction) into the volume over the 10G ethernet.

If I create a RWO volume on the same sata storage class it easily gets 250 - 300MB/s write speed but I want the volume avaible to many worker pods though, so I can't just go with RWO.  I'm using the default NFS options from the code but changing NFS version to 4.2.  

I feel like there has to be a reason why the performance is so bad, it can't be this slow.

And then a separate issue is that if I stop writing and try to just use some of the stored data in several pods...  if I read using too many pods at once, LH will start marking the replica as failed, and trying to delete it and rebuild it and stuff... it even led to corrupt filesystem...and the share manager was crash looping. That is highly undesirable because this is a very large volume.  The delete (really looks like its a merge of snapshots) can take hours or days... and then it has to do a rebuild which then takes hours or days also even though there was probably nothing wrong with the data at all and I would have just needed to kill off the out of control job that was creating too many connections at once and probably was why it thought the replica was down.

So my questions for discussion are:

1) Is this really slow RWX performance expected? and if not do you have experience with what NFS options would fix it?  It would be wonderful if there was a way to experiment with the client/server options without having to make whole new storage classes and volumes since that is such a LONG process with large volumes (i could test with small ones that can create and destroy faster but totally blocks creating the final desired one until done tinkering). 

2) Is there anything that can be done to prevent automatic actions that destroy replicas and try to recreate them without regard to whether it would have really been warranted or not.

**Comment by derekbit** [comment]
We don't recommend Using spinning disks (HDD) as underlying disks because
  - Easy to hit the IO starvation and lead to rebuild failed
  - Slow replica rebuilding

For IO intensive environment, SSDs are recommended. Please see the [best practices](https://longhorn.io/docs/1.6.0/best-practices/).

For the NFS protocol, I think the protocol can introduce performance drop, but we don't have numbers. We will work on it as well as update the document.

**Reply by linuxfreakus** [reply]
For the application I have, the data is large but very slowly changing... investing $100k or more in SSD/NVMe storage for that doesn't make sense.  There should not be any I/O starvation with only a single process running rsync.... yes it might be slow but that would be acceptable if the speed was reasonably comparable to what the RWO volume can get.

If there are ways to alter the behavior so that it doesn't try to move or rebuild replicas that aren't really failed that might help in this situation as well.  I see there are a couple params I can tweak to give a little more time before it thinks a replica is failed but nothing lets me totally require approval before taking extremely destructive and disruptive actions (that I can see anyway)

**Comment by linuxfreakus** [comment]
Another datapoint: I made a pod which mounted a straight up NFS share running on the k8s host node directly (not controlled by Longhorn). NFS 4.2 with the same set of options used in Longhorn.... and I'm easily writing 250 - 280 MB/s to that share.

The exact same filesystem that the Longhorn RWX is on, using LVM on spinning sata.

The host is ubuntu 22.04 just using the vanilla nfs-kernel-server package from apt. No customizations.

So it would appear that there is definitely something funky going on with Longhorn's use of NFS in the RWX volumes.

Not sure if there are any changes in 1.6.0 that would be expected to make a difference... I could try that on our dev environment perhaps, but I'm not changing prod right now.

**Comment by echo-devnull** [comment]
We are having the same issues.

I have tried RWX volumes on NVME and SSD type disks. The load jumps to 80 on a AMD 4750G type processor and 64GB of memory.

Same disks with a RWO volume is zippy fast. Native NVME speeds I'd say. But as soon as we use the RWX and ganesha nfs pods are used: Extreme iowait, Extreme load, super slow speeds. Even when only 1 pod is using this RWX volume.

We can replicate this every time by just setting up a RWX volume.

We tried with XFS formatted drives, EXT4, 1 replicated volume, 2, 3, does not matter. RWX is always extremely slow and basically unusable at this point.


(Debian 12, RKE2 with 1.28.11, longhorn 1.7.0)

**Reply by PhanLe1010** [reply]
@echo-devnull Could you please help us by creating a ticket with detailed reproducing steps at https://github.com/longhorn/longhorn/issues/new/choose? We would like to try reproducing it

cc @james-munson

**Comment by derekbit** [comment]
cc @james-munson

**Comment by Szwendacz99** [comment]
I am trying to copy about 1.5TB of data, including big files to RWX volume on HDD, and it basically always crashes, either with ugly dmesg messages about corrupted ext4 filesystem, or by timeouts. Adding some flags, like sync to nfs mount options seem to help with stability somewhat, but the speed is obviously even more degraded.

**Reply by PhanLe1010** [reply]
@Szwendacz99 How many replica does the RWX volume have? 

RWX might be slow but shouldn't crash under heavy load.
Can you change this to setting to bigger value to see if it helps to avoid the crashing?

1. Engine Replica Timeout https://longhorn.io/docs/1.7.1/references/settings/#engine-replica-timeout
2. Timeout of HTTP Client to Replica File Sync Server https://longhorn.io/docs/1.7.1/references/settings/#timeout-of-http-client-to-replica-file-sync-server

**Reply by Szwendacz99** [reply]
My k8s is single node currently, so no replicas. Currently the HDD is with LUKS+btrfs with copyonwrite and compression enabled, but for short time I tested behavior of plain xfs partition instead, and suprisingly on plain xfs it failed even faster.

**Reply by PhanLe1010** [reply]
@Szwendacz99 

Even on a single node, the volume still have a local replica so the setting Engine Replica Timeout https://longhorn.io/docs/1.7.1/references/settings/#engine-replica-timeout might still affect it.

Can you take a look at the logs of longhorn-manager, instance-manager, and share-manager pods when the crash happen to see if we have more clue?

**Comment by jasonmoore-gdms** [comment]
> if I read using too many pods at once, LH will start marking the replica as failed, and trying to delete it and rebuild it and stuff... it even led to corrupt filesystem...and the share manager was crash looping

We also experienced issues with share-manager crashing and nfs daemon not ready.  We support a customer design where multiple replicasets are using the same shared Longhorn RWX volume.  This is not a design we recommend, but found that LH can support it after upgrading to v1.6.3

Specifically:
 [v1.6.2][IMPROVEMENT] Do not terminate nfs-ganesha in share-manager pod after failing to access recovery backend [8346](https://github.com/longhorn/longhorn/issues/8346) - [@derekbit](https://github.com/derekbit) [@chriscchien](https://github.com/chriscchien)
